build.assets/tooling/cmd/changelog/main.go

```diff
@@ -0,0 +1,222 @@
+func getBranch(branch, dir string) (string, error) {
+	if branch == "" { // flag and env var not set, attempt to find
+		// get ref
+		ref, err := git.RunCmd(dir, "symbolic-ref", "HEAD")
+		if err != nil {
+			return "", trace.Wrap(err, "not on a branch")
+		}
+
+		// remove prefix and ensure that branch is in expected format
+		branch, _ = strings.CutPrefix(ref, "refs/heads/")
+		if branch == ref {
+			return "", trace.Errorf("not on a branch: %s", ref)
+		}
+
+		// if the branch is not in the branch/v* format then check it's root
+		if !strings.HasPrefix(branch, "branch/v") {
+			fbranch, err := getForkedBranch(dir)
+			if err != nil {
+				return "", trace.Wrap(err, "could not determine a root branch")
+			}
+			branch = fbranch
+		}
+	}
+
+	if !strings.HasPrefix(branch, "branch/v") {
+		return "", trace.Errorf("not on a release branch, expected 'branch/v*', got %s", branch)
+	}
+
+	return branch, nil
+}
+
+// makePrintVersion will run 'make -s print-version'
+func makePrintVersion(dir string) (string, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command("make", "-s", "print-version")
+	cmd.Dir = dir
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
+
+	err := cmd.Run()
+
+	if err != nil {
+		return strings.TrimSpace(stderr.String()), trace.Wrap(err, "can't get last released version")
+	}
+	out := strings.TrimSpace(stdout.String())
+
+	return "v" + out, nil
+}
+
+func getLastVersion(baseTag, dir string) (string, error) {
+	if baseTag != "" {
+		return baseTag, nil
+	}
+
+	// get root dir of repo
+	topDir, err := git.RunCmd(dir, "rev-parse", "--show-toplevel")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	lastVersion, err := makePrintVersion(topDir)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return lastVersion, nil
+}
+
+func getTimestamps(dir string, entDir string, lastVersion string) (lastRelease, lastEnterpriseRelease, lastEnterpriseModify string, err error) {
+	// get timestamp since last release
+	since, err := git.RunCmd(dir, "show", "-s", "--date=format:%Y-%m-%dT%H:%M:%S%z", "--format=%cd", lastVersion)
+	if err != nil {
+		return "", "", "", trace.Wrap(err, "can't get timestamp of last release")
+	}
+	// get timestamp of last enterprise release
+	sinceEnt, err := git.RunCmd(dir, "-C", "e", "show", "-s", "--date=format:%Y-%m-%dT%H:%M:%S%z", "--format=%cd", lastVersion)
+	if err != nil {
+		return "", "", "", trace.Wrap(err, "can't get timestamp of last enterprise release")
+	}
+	// get timestamp of last commit of enterprise
+	entTime, err := git.RunCmd(entDir, "log", "-n", "1", "--date=format:%Y-%m-%dT%H:%M:%S%z", "--format=%cd")
+	if err != nil {
+		return "", "", "", trace.Wrap(err, "can't get last modified time of e")
+	}
+	return since, sinceEnt, entTime, nil
+}
+
+func prereqCheck() error {
+	if err := git.IsAvailable(); err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := exec.LookPath("git"); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func main() {
```

Comment:
Nit: This is a minor comment/nitpick but do you mind rearranging functions so "main" is on top and the functions it calls are defined roughly in the order they are being called? This way you can follow the code reading from top to bottom like a book - right now it starts with some utility functions so by the time I get to "main" I already forgot what I read before :)

Reply:
Oh no worries. My brain switched it up. I had it almost exactly the opposite haha. It should be easier to follow now.